relay-replays/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+//! They are ingested incrementally and are available for viewing after the first two segments
+//! (segment 0 and 1) have completed ingestion.
+//!
+//! # Protocol
+//!
+//! Relay is expecting a JSON object with some mandatory metadata.  However, environment and user
+//! metadata is usually sent in addition to the minimal payload.
+//!
+//! ```json
+//! {
+//!     "type": "replay_event",
+//!     "replay_id": "d2132d31b39445f1938d7e21b6bf0ec4",
+//!     "event_id": "63c5b0f895441a94340183c5f1e74cd4",
+//!     "segment_id": 0,
+//!     "timestamp": 1597976392.6542819,
+//!     "replay_start_timestamp": 1597976392.6542819,
+//!     "urls": ["https://sentry.io"],
+//!     "error_ids": ["d2132d31b39445f1938d7e21b6bf0ec4"],
+//!     "trace_ids": ["63c5b0f895441a94340183c5f1e74cd4"],
+//!     "requests": {
+//!         "headers": {"User-Agent": "Mozilla/5.0..."}
+//!     },
+//! }
+//! ```
+use std::collections::HashMap;
+use std::fmt::Write;
+use std::net::IpAddr;
+
+use serde::{Deserialize, Serialize};
+use serde_json::Error;
+
+use relay_general::user_agent::{parse_device, parse_os, parse_user_agent, Device};
+
+pub fn normalize_replay_event(
+    replay_bytes: &[u8],
+    detected_ip_address: Option<IpAddr>,
+) -> Result<Vec<u8>, Error> {
+    //! Validates and accepts a replay-event item as input and returns modified output.
+    //!
+    //! The output of this function will return a valid replay-event payload with the addition of
+    //! browser, device, and operating-system metadata.  The output will also contain a default
+    //! ip-address value for the user if none was provided.
+    let mut replay_input: ReplayInput = serde_json::from_slice(replay_bytes)?;
+
+    // Set user-agent metadata from requests object.
+    replay_input.set_user_agent_meta();
+
+    // Set user ip-address if needed.
+    match detected_ip_address {
+        Some(ip_address) => replay_input.set_user_ip_address(ip_address),
+        None => (),
+    }
+
+    serde_json::to_vec(&replay_input)
+}
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+struct ReplayInput {
+    #[serde(rename = "type")]
+    ty: String,
+    replay_id: String,
+    event_id: String,
+    segment_id: u8,
+    timestamp: f64,
+    replay_start_timestamp: Option<f64>,
+    contexts: Option<Contexts>,
+    dist: Option<String>,
+    platform: Option<String>,
+    environment: Option<String>,
+    release: Option<String>,
+    tags: Option<HashMap<String, String>>,
+    sdk: Option<VersionedMeta>,
+    #[serde(default)]
+    urls: Vec<String>,
+    #[serde(default)]
+    error_ids: Vec<String>,
+    #[serde(default)]
+    trace_ids: Vec<String>,
+    #[serde(default)]
+    user: User,
+    #[serde(default)]
+    requests: Requests,
+}
+
+impl ReplayInput {
+    fn set_user_agent_meta(&mut self) {
+        let user_agent = &self.requests.headers.user_agent;
+
+        let device = parse_device(user_agent);
+
+        let ua = parse_user_agent(user_agent);
```

Comment:
@jan-auer not sure how strict we are about this convention, but I think we qualify imports from sibling crates as `user_agent::parse_user_agent(...)`, correct?

Reply:
Went ahead and implemented that way regardless.  Should add clarity to the origin of the function.